File: src/Bionic.Bridge.Capacitor/AppBridge.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.JSInterop;

namespace Bionic.Bridge.Capacitor {
    public static class AppBridge {
        private static Dictionary<string, Action<AppState>> stateWatchers = new Dictionary<string, Action<AppState>>();

        private static Dictionary<string, Action<AppUrlOpen>> urlOpenWatchers =
            new Dictionary<string, Action<AppUrlOpen>>();

        private static Dictionary<string, Action<AppRestoredResult>> restoredResultWatchers =
            new Dictionary<string, Action<AppRestoredResult>>();

        private static Dictionary<string, Action<AppUrlOpen>> backButtonWatchers =
            new Dictionary<string, Action<AppUrlOpen>>();

        private const string AppCanOpenUrlFunctionName = "BionicBridge.Capacitor.App.canOpenUrl";
        private const string AppExitFunctionName = "BionicBridge.Capacitor.App.exitApp";
        private const string AppGetLaunchUrlFunctionName = "BionicBridge.Capacitor.App.getLaunchUrl";
        private const string AppOpenUrlFunctionName = "BionicBridge.Capacitor.App.openUrl";
        private const string AppStateChangeFunctionName = "BionicBridge.Capacitor.App.addAppStateChangeListener";
        private const string AppAppUrlOpenFunctionName = "BionicBridge.Capacitor.App.addAppUrlOpenListener";
        private const string AppRestoredResultFunctionName = "BionicBridge.Capacitor.App.addAppRestoredResultListener";
        private const string AppBackButtonFunctionName = "BionicBridge.Capacitor.App.addBackButtonListener";
        private const string AppRemoveListenerFunctionName = "BionicBridge.Capacitor.App.removeListener";

        public static Task<bool> CanOpenUrl(string url) =>
            JSRuntime.Current.InvokeAsync<bool>(AppCanOpenUrlFunctionName, url);

        public static Task ExitApp() =>
            JSRuntime.Current.InvokeAsync<object>(AppExitFunctionName);

        public static Task<string> GetLaunchUrl() =>
            JSRuntime.Current.InvokeAsync<string>(AppGetLaunchUrlFunctionName);

        public static Task<bool> OpenUrl(string url) =>
            JSRuntime.Current.InvokeAsync<bool>(AppOpenUrlFunctionName, url);

        public static async Task AddAppStateChangeListener(string id, Action<AppState> callback) {
            await JSRuntime.Current.InvokeAsync<object>(AppStateChangeFunctionName, id);
            stateWatchers.Add(id, callback);
        }

        public static async Task AddAppUrlOpenListener(string id, Action<AppUrlOpen> callback) {
            await JSRuntime.Current.InvokeAsync<object>(AppAppUrlOpenFunctionName, id);
            urlOpenWatchers.Add(id, callback);
        }

        public static async Task AddAppRestoredResultListener(string id, Action<AppRestoredResult> callback) {
            await JSRuntime.Current.InvokeAsync<object>(AppRestoredResultFunctionName, id);
            restoredResultWatchers.Add(id, callback);
        }

        public static async Task AddAppBackButtonListener(string id, Action<AppUrlOpen> callback) {
            await JSRuntime.Current.InvokeAsync<object>(AppBackButtonFunctionName, id);
            backButtonWatchers.Add(id, callback);
        }

        public static Task RemoveListener(string id) {
            if (stateWatchers.ContainsKey(id)) stateWatchers.Remove(id);
            else if (urlOpenWatchers.ContainsKey(id)) urlOpenWatchers.Remove(id);
            else if (restoredResultWatchers.ContainsKey(id)) restoredResultWatchers.Remove(id);
            else if (backButtonWatchers.ContainsKey(id)) backButtonWatchers.Remove(id);
            return JSRuntime.Current.InvokeAsync<object>(AppRemoveListenerFunctionName, id);
        }

        [JSInvokable]
        public static void ReportAppStateChange(string id, AppState state) =>
            stateWatchers[id](state);

        [JSInvokable]
        public static void ReportAppUrlOpen(string id, AppUrlOpen data) =>
            urlOpenWatchers[id](data);

        [JSInvokable]
        public static void ReportAppRestoredResult(string id, AppRestoredResult data) =>
            restoredResultWatchers[id](data);

        [JSInvokable]
        public static void ReportBackButton(string id, AppUrlOpen data) =>
            backButtonWatchers[id](data);
    }

    public class AppState {
        public bool isActive;
    }

    public class AppUrlOpen {
        public string url;
        public object iosSourceApplication;
        public bool iosOpenInPlace;
    }

    public class AppRestoredResult {
        public string pluginId;
        public string methodName;
        public string data;
    }
}

File: src/Bionic.Bridge.Capacitor.JS/src/AppBridge.ts
import {AppRestoredResult, AppState, AppUrlOpen, PluginListenerHandle, Plugins} from "@capacitor/core";

const {App} = Plugins;

class Watcher {
  id: string | null = null;
  handler: PluginListenerHandle | null = null;

  public remove(): void {
    if (this.handler) {
      this.handler.remove();
      this.handler = null;
    }
  }

  public appStateChangeListener(state: AppState): void {
    // @ts-ignore
    DotNet.invokeMethodAsync('BionicBridgeCapacitor', 'ReportAppStateChange', this.id, state);
  }

  public appUrlOpenListener(data: AppUrlOpen): void {
    // @ts-ignore
    DotNet.invokeMethodAsync('BionicBridgeCapacitor', 'ReportAppUrlOpen', this.id, data);
  }

  public appRestoredResultListener(data: AppRestoredResult): void {
    // @ts-ignore
    DotNet.invokeMethodAsync('BionicBridgeCapacitor', 'ReportAppRestoredResult', this.id, data);
  }

  public backButtonListener(data: AppUrlOpen): void {
    // @ts-ignore
    DotNet.invokeMethodAsync('BionicBridgeCapacitor', 'ReportBackButton', this.id, data);
  }
}

interface IApp {
  canOpenUrl(url: string): Promise<boolean>;

  exitApp(): Promise<void>;

  getLaunchUrl(): Promise<string | null>;

  openUrl(url: string): Promise<boolean>;

  addAppStateChangeListener(id: string): Promise<void>;

  addAppUrlOpenListener(id: string): Promise<void>;

  addAppRestoredResultListener(id: string): Promise<void>;

  addBackButtonListener(id: string): Promise<void>;

  removeListener(id: string): Promise<void>;
}

export class AppBridge implements IApp {
  private static watchers = new Map<string, Watcher>();

  async canOpenUrl(url: string): Promise<boolean> {
    if (!App) return Promise.reject("App not available.");
    const result = await App.canOpenUrl({url});
    return result ? result.value : false;
  }

  async exitApp(): Promise<void> {
    if (!App) return Promise.reject("App not available.");
    await App.exitApp();
  }

  async getLaunchUrl(): Promise<string | null> {
    if (!App) return Promise.reject("App not available.");
    const result = await App.getLaunchUrl();
    return result && result.url ? result.url : null;
  }

  async openUrl(url: string): Promise<boolean> {
    if (!App) return Promise.reject("App not available.");
    const result = await App.openUrl({url});
    return result ? result.completed : false;
  }

  async addAppStateChangeListener(id: string): Promise<void> {
    if (!App) return Promise.reject<void>("App not available.");

    try {
      const watcher = new Watcher();
      watcher.id = id;
      watcher.handler = await App.addListener("appStateChange", watcher.appStateChangeListener.bind(watcher));
      AppBridge.watchers.set(id, watcher);
      return new Promise<void>(resolve => resolve());
    } catch (e) {
      return new Promise<void>((r, reject) => reject(e.message));
    }
  }

  async addAppUrlOpenListener(id: string): Promise<void> {
    if (!App) return Promise.reject<void>("App not available.");

    try {
      const watcher = new Watcher();
      watcher.id = id;
      watcher.handler = await App.addListener("appUrlOpen", watcher.appUrlOpenListener.bind(watcher));
      AppBridge.watchers.set(id, watcher);
      return new Promise<void>(resolve => resolve());
    } catch (e) {
      return new Promise<void>((r, reject) => reject(e.message));
    }
  }

  async addAppRestoredResultListener(id: string): Promise<void> {
    if (!App) return Promise.reject<void>("App not available.");

    try {
      const watcher = new Watcher();
      watcher.id = id;
      watcher.handler = await App.addListener("appRestoredResult", watcher.appRestoredResultListener.bind(watcher));
      AppBridge.watchers.set(id, watcher);
      return new Promise<void>(resolve => resolve());
    } catch (e) {
      return new Promise<void>((r, reject) => reject(e.message));
    }
  }

  async addBackButtonListener(id: string): Promise<void> {
    if (!App) return Promise.reject<void>("App not available.");

    try {
      const watcher = new Watcher();
      watcher.id = id;
      watcher.handler = await App.addListener("backButton", watcher.backButtonListener.bind(watcher));
      AppBridge.watchers.set(id, watcher);
      return new Promise<void>(resolve => resolve());
    } catch (e) {
      return new Promise<void>((r, reject) => reject(e.message));
    }
  }

  async removeListener(id: string): Promise<void> {
    try {
      const watcher = AppBridge.watchers.get(id);
      if (watcher && watcher.handler) {
        watcher.remove();
        AppBridge.watchers.delete(id);
      }
      return new Promise<void>(resolve => resolve());
    } catch (e) {
      return new Promise<void>((r, reject) => reject(e.message));
    }
  }
}
